quinn-proto/src/cid_generator.rs

```diff
@@ -0,0 +1,60 @@
+use rand::RngCore;
+
+use crate::shared::ConnectionId;
+use crate::MAX_CID_SIZE;
+
+/// Generates connection IDs for incoming connections
+pub trait ConnectionIdGenerator: Send {
+    /// Generates a new CID
+    ///
+    /// Connection IDs MUST NOT contain any information that can be used by
+    //    an external observer (that is, one that does not cooperate with the
+    //    issuer) to correlate them with other connection IDs for the same
+    //    connection.
+    fn generate_cid(&mut self) -> ConnectionId;
+    /// Performs any validation if it is needed (e.g. HMAC, etc)
+    ///
+    /// Apply validation check on those CIDs that may still exist in hash table
+    ///   but considered invalid by application-layer logic.
+    /// e.g., We may want to limit the amount of time for which a CID is valid
+    ///   in order to reduce the number of valid IDs that could be accumulated
+    ///   by an attacker.
+    fn validate_cid(&mut self, _cid: &ConnectionId) -> bool {
+        true
+    }
```

Comment:
Good call on the default implementation! I'd still like to understand the problem this is solving in more detail, though. Quinn already limits the number of connection IDs that may be outstanding per connection via the mechanisms discussed in [Issuing Connection IDs](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#name-issuing-connection-ids), enforced [inside `CidQueue`](https://github.com/djc/quinn/blob/master/quinn-proto/src/cid_queue.rs#L41-L43), currently hardcoded to at most 4. This is communicated to the peer via the `active_connection_id_limit` transport parameter, which helps ensure that well-behaved peers don't encounter issues.

Is there a CID exhaustion attack that this does not cover? If so, should we pursue a mechanism that also cleans them out of the hash table proactively, so as to avoid memory exhaustion?

Reply:
> Is there a CID exhaustion attack that this does not cover? 

i agree that for well-behaved peers we shouldn't expect issues. quinn has followed RFC to protect each pair of communications.

This `fn validate_cid`  is a precaution to avoid when an attack can somehow eardrop `multiple` pairs of communications and accumulate many CIDs if server doesn't proactively invalidate them. As a result, I will suggest peer (especially server side) can push `NEW_CONNECTION_ID` with updated `retire_prior_to` field  (currently, quinn always set it to 0) to rotate CID. This will somehow reduce the risk of "CID exhaustion attack". I have already finished the code on my local branch.

A well-behaved peer should reply with `RETIRE_CONNECTION_ID` to complete the CID rotation. If not,  `validate_cid` then will kick in to stop the server to keep using the retired ones. 

>If so, should we pursue a mechanism that also cleans them out of the hash table proactively, so as to avoid memory exhaustion?

I think it will be good if we can clean those CIDs that are invalidated. Could be a future improvement. 👍

Reply:
Oh, I see, the real goal here is CID key rotation or similar. That's definitely something we want to support. My concern is that the approach currently taken here will place well-behaved peers at risk of connection loss due to not employing `retire_prior_to`. Additionally, there's no mechanism introduced to issue new CIDs, so a long-lived connection will run out and be lost. Therefore I don't think `validate_cid` is useful as written.

Since it sounds like you have a better approach almost ready to go, can we `validate_cid` here, merge the rest, and address CID rotation in a future PR?

Reply:
> Since it sounds like you have a better approach almost ready to go, can we omit validate_cid here, merge the rest, and address CID rotation in a future PR?

Wouldn't this break users if the start implementing the new trait, and the additional method has to be added later on?
If it's forseeable that it will be required or even that it could be useful there shouldn't be a lot of harm with adding it - even if it's not called right now. The implementation should be straightforward for most CID formats.

Reply:
> My concern is that the approach currently taken here will place well-behaved peers at risk of connection loss due to not employing retire_prior_to

Totally understood. I double check the description in [RFC](https://www.ietf.org/id/draft-ietf-quic-transport-31.html#name-consuming-and-retiring-conn) where I quote 
> "Upon receipt of an increased Retire Prior To field, the peer MUST stop using the corresponding connection IDs and retire them with RETIRE_CONNECTION_ID frames before adding the newly provided connection ID to the set of active connection IDs."

Since peer `MUST` has a reply, I don't think if a peer who doesn't employ the `retire_prior_to` can be categorized as "well-behaved" (at least it is not RFC-compliant). 

The `validate_cid` function provides a API/approach to apply what is mentioned in RFC where
> "The endpoint SHOULD continue to accept the previously issued connection IDs until they are retired by the peer. If the endpoint can no longer process the indicated connection IDs, it MAY close the connection."

Since it is a `MAY` in RFC, the CID generator creator can thus decide whether/when we want to close the connection based on their own implementation. In the current PR, the false result of `validate_cid` will only ignore the packets from peer, which is not exact equivalent to closing the connection. This could be a future improvement.

With that said, 
> Since it sounds like you have a better approach almost ready to go

I have a PR ready that implements the advance of `retire_prior_to` but the not form of cid generator trait.

Since the current calling of `validate_cid` is not perfect yet. I am willing to rollback this API. Regarding to Matthias's concern, it is also feasible to add it back in the future without breaking compatibility if we provide a default implementation as I have now: users can update their implementation if they want to use this feature.

Reply:
> Since peer MUST has a reply, I don't think if a peer who doesn't employ the retire_prior_to can be categorized as "well-behaved" (at least it is not RFC-compliant).

Right, the problem is that quinn itself does not yet generate `retire_prior_to`, hence if any implementer ever returned `false` from `validate_cid` under current conditions it would necessarily cause a CID that might still be legally used by the peer to be ignored.

> If it's forseeable that it will be required or even that it could be useful there shouldn't be a lot of harm with adding it

Ideally we'd want to tightly couple the local invalidation of CIDs with the issuing of a `retire_prior_to` and (deferred) updates to the Endpoint's hash table so that everything is guaranteed to stay in sync, which will likely require a different interface.